src/Telomare/Eval.hs

```diff
@@ -221,42 +242,58 @@ schemeEval iexpr = do
   scheme <- hGetContents mhout
   putStrLn scheme
 
+evalLoopCore :: IExpr
+             -> (String -> String -> IO String)
+             -> String
+             -> [String]
+             -> IO String
+evalLoopCore iexpr accumFn initAcc manualInput =
+  let wrappedEval = funWrap eval iexpr
+      mainLoop :: String -> [String] -> Maybe (String, IExpr) -> IO String
+      mainLoop acc strInput s = do
+        (out, nextState) <- wrappedEval s
+        newAcc <- accumFn acc out
+        case nextState of
+          Nothing -> pure acc
+          Just Zero -> pure $ newAcc <> "\n" <> "done"
+          Just ns -> do
+            (inp, rest) <-
+              if null strInput
+              then (, []) <$> getLine
+              else pure (head strInput, tail strInput)
+            mainLoop newAcc rest $ pure (inp, ns)
+  in mainLoop initAcc manualInput Nothing
 
 evalLoop :: IExpr -> IO ()
-evalLoop iexpr =
-  let mainLoop s = do
-        result <- simpleEval $ app iexpr s
-        case result of
-          Zero -> putStrLn "aborted"
-          (Pair disp newState) -> do
-            putStrLn . g2s $ disp
-            case newState of
-              Zero -> putStrLn "done"
-              _ -> do
-                inp <- s2g <$> getLine
-                mainLoop $ Pair inp newState
-          r -> putStrLn ("runtime error, dumped " <> show r)
-  in mainLoop Zero
+evalLoop iexpr = void $ evalLoopCore iexpr printAcc "" []
+  where
+    printAcc _ out = do
+      putStrLn out
+      pure ""
+
+evalLoopWithInput :: [String] -> IExpr -> IO String
+evalLoopWithInput inputList iexpr = evalLoopCore iexpr printAcc "" inputList
+  where
+    printAcc acc out = pure (acc <> "\n" <> out)
```

Comment:
ah, looks like this is the newline issue. It's not a big deal, though

Reply:
I already notified you in a DM, but for completeness I'll repeat it here:

So this change in `printAcc` makes `cabal test` stall:

```haskell
evalLoopWithInput :: [String] -> IExpr -> IO String
evalLoopWithInput inputList iexpr = evalLoopCore iexpr printAcc "" inputList
  where
    -- printAcc acc out = pure (acc <> "\n" <> out)
    printAcc acc out = if null acc && null out
                       then pure "\n"
                       else if null acc
                            then pure out
                            else pure $ acc <> "\n" <> out
```

The stalling occurs with a simpler `printAcc` too:

```haskell
printAcc acc out = if null acc
                   then pure out
                   else pure (acc <> "\n" <> out)

```

my curiosity has sadly made me think of this more than I thought I would. lol..

Reply:
The stalling was a symptom to a bigger problem: there was a problem in the `shrink` of `Pattern` (a type that's parsed form the case-patterns to be matched.

Still not sure why changing `printAcc` makes that problem evident (or why it did work with the not so smart version `pure $ acc <> "\n" <> out`).

I also noticed that in my testing `@?=` is used and that it isn't a symmetric operator, the left part is the thing to be tested and the right part is the expected value. This next snippet shows why it matters (expected is inverted with result):

```
  Unit tests on case expressions
    test case with int leaves:          FAIL (0.04s)
      test/CaseTests.hs:73:
      expected: "\nTrue\ndone"
       but got: "True\ndone"
```